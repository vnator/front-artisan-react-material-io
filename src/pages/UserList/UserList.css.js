import { makeStyles } from '@material-ui/core';

const useStyles = makeStyles(({ palette, spacing }) => ({
  userList: {},
  title: {
    marginBottom: spacing(4),
  },
}));

export { useStyles };
