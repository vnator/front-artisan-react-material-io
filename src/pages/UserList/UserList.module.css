.UserList {
  --border: 1px solid var(--white-glass);
  --cell: 260px;
  --size: calc((var(--cell) * 2) + var(--space-large));

  color: var(--white-full);
  min-height: 100vh;
  height: auto;
  max-width: var(--size);
  margin: 0 auto;
}

.list {
  display: flex;
  flex-direction: column;
  text-align: left;
  width: auto;
  margin: 0 auto;
  width: var(--size);
  padding: 0;
}

.item,
.head,
.user {
  display: flex;
  box-sizing: border-box;
}

.head {
  padding: var(--space-base);
}

.head .iconButton {
  min-width: var(--cell);
}

.head .icon {
  margin: 0 var(--space-base);
}

.iconButton {
  text-align: left;
  border: 0;
  background: none;
  color: var(--white-full);
  text-transform: uppercase;
  padding: 0;
  font-weight: bold;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.icon {
  width: 22px;
  height: 22px;
  fill: var(--first);
  transition: var(--transition-base);
}

.iconButton:hover > .icon {
  fill: var(--white-full);
}

.item {
  border: var(--border);
  border-bottom: 0;
}

.list > .item:nth-last-child(2) {
  border-bottom: var(--border);
}

.user {
  cursor: pointer;
  width: 100%;
  color: var(--white-half);
  background: none;
  border: 0;
  text-align: left;
  padding: var(--space-base);
  transition: var(--transition-base0);
}

.user:hover {
  background: var(--grey-dark);
}

.cell,
.order {
  width: var(--cell);
}

.foot {
  display: flex;
  justify-content: flex-end;
  list-style: none;
  padding: 0;
  height: 30px;
}

.page {
  width: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
}
