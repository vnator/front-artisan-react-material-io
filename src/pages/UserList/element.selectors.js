const el = Object.freeze({
  btnSortByName: 'test-user_list-order_by_name',
  btnSortByEmail: 'test-user_list-order_by_email',
  btnNextPage: 'test-user_list-next_page',
  btnPrevPage: 'test-user_list-prev_page',
  btnGetUser: 'test-user_list-get_user',
  btnNewUser: 'test-user_list-new_user',
});

export { el };
