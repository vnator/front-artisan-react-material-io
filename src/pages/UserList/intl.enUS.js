const userList_enUS = {
  title: 'GraphQL Users Request',
  newUser: 'add new user',
  list: {
    name: 'name',
    email: 'email',
  },
};

export { userList_enUS };
