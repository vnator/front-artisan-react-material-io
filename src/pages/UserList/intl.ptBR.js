const userList_ptBR = {
  title: 'GraphQL Users Request',
  newUser: 'add new user',
  list: {
    name: 'name',
    email: 'email',
  },
};

export { userList_ptBR };
