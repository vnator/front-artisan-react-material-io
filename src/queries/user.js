import gql from 'graphql-tag';

const USER_FRAGMENTS = Object.freeze({
  USER_FIELDS: gql`
    fragment userFields on User {
      id
      name
      email
      dateOfBirth
      gender
    }
  `,
});

const USER = {
  GET_USER: gql`
    query User($id: ID) {
      user(id: $id) {
        ...userFields
      }
    }

    ${USER_FRAGMENTS.USER_FIELDS}
  `,
  GET_USER_LIST: gql`
    query Users($sortField: SortFieldUser, $skip: Int, $limit: Int) {
      users(sortField: $sortField, skip: $skip, limit: $limit) {
        ...userFields
      }
    }

    ${USER_FRAGMENTS.USER_FIELDS}
  `,
  UPSERT_USER: gql`
    mutation UpsertUser(
      $name: String
      $email: Email
      $dateOfBirth: Date
      $gender: Gender
    ) {
      upsertUser(
        name: $name
        email: $email
        dateOfBirth: $dateOfBirth
        gender: $gender
      ) {
        ...userFields
      }
    }

    ${USER_FRAGMENTS.USER_FIELDS}
  `,
};

export { USER, USER_FRAGMENTS };
