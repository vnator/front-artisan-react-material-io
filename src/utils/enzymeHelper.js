import { IntlProvider, createIntl } from 'react-intl';
import { mount, shallow } from 'enzyme';
import { flattenMessages } from '../config/flattenMessages';
import { messages } from '../messages';

// You can pass your messages to the IntlProvider. Optional: remove if unneeded.

const defaultLocale = 'pt-BR';
const locale = defaultLocale;
const wrapMessages = flattenMessages(messages[locale]);

function mountWithIntl(node) {
  return mount(node, {
    wrappingComponent: IntlProvider,
    wrappingComponentProps: {
      locale,
      defaultLocale,
      messages: wrapMessages,
    },
  });
}

function shallowWithIntl(node) {
  return shallow(node, {
    wrappingComponent: IntlProvider,
    wrappingComponentProps: {
      locale,
      defaultLocale,
      messages: wrapMessages,
    },
  });
}

const intl = createIntl({
  locale: 'pt-BR',
  defaultLocale: 'pt-BR',
  messages: wrapMessages,
});

export { mountWithIntl, shallowWithIntl, intl };
