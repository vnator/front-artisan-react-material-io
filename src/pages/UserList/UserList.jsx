import React from 'react';
import Type from 'prop-types';

import { useHistory } from 'react-router-dom';
import { useQuery } from '@apollo/react-hooks';
import { useIntl } from 'react-intl';

import {
  ArrowDropDown,
  ArrowDropUp,
  ArrowLeft,
  ArrowRight,
  EditOutlined,
} from '@material-ui/icons';

import { MAIN_ROUTES } from '../../const/routes';
import { SORT_FIELD } from '../../const/userListParams';
import { USER } from '../../queries/user';
import { usePagination } from './usePagination';
import { useStyles } from './UserList.css';
import {
  Typography,
  Button,
  IconButton,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
  Tooltip,
} from '@material-ui/core';

import {
  Table,
  TabContainer,
  Head,
  Footer,
  FooterRow,
  Cell,
} from '../../components/Table.css';

const HeadCell = ({ field, sortField, event }) => {
  const { formatMessage } = useIntl();

  return (
    <Head>
      <Button
        onClick={() => event(field)}
        endIcon={sortField === field ? <ArrowDropDown /> : <ArrowDropUp />}>
        {formatMessage({
          id: `users.list.${field}`,
        })}
      </Button>
    </Head>
  );
};

const UserList = () => {
  const history = useHistory();
  const classes = useStyles();
  const [pag, { nextPage, prevPage, orderBy }] = usePagination(SORT_FIELD.NAME);

  const { formatMessage } = useIntl();

  const { data } = useQuery(USER.GET_USER_LIST, {
    variables: pag.params,
  });

  return (
    <div>
      <Typography color="primary" variant="h4" className={classes.title}>
        {formatMessage({
          id: 'users.title',
        })}
      </Typography>

      <TabContainer>
        <Table>
          <TableHead>
            <TableRow>
              {Object.values(SORT_FIELD).map((val) => (
                <HeadCell
                  field={val}
                  sortField={pag.params.sortField}
                  event={orderBy}
                />
              ))}
              <Head size="small" />
            </TableRow>
          </TableHead>
          <TableBody>
            {data &&
              data.users.map((user) => (
                <TableRow key={user.id}>
                  <TableCell size="small">{user.name}</TableCell>
                  <TableCell size="small">{user.email}</TableCell>
                  <Cell size="small">
                    <Tooltip
                      title={formatMessage({ id: 'users.edit' })}
                      aria-label={formatMessage({ id: 'users.edit' })}>
                      <IconButton
                        onClick={() => history.push(MAIN_ROUTES.USER(user.id))}>
                        <EditOutlined />
                      </IconButton>
                    </Tooltip>
                  </Cell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TabContainer>
      <Footer>
        <FooterRow>
          <div>
            <IconButton onClick={prevPage}>
              <ArrowLeft />
            </IconButton>
            <strong>{pag.page}</strong>
            <IconButton onClick={nextPage}>
              <ArrowRight />
            </IconButton>
          </div>

          <div className={classes.newUser}>
            <Button
              size="large"
              color="primary"
              onClick={() => history.push(MAIN_ROUTES.USER(0))}>
              {formatMessage({
                id: 'users.newUser',
              })}
            </Button>
          </div>
        </FooterRow>
      </Footer>
    </div>
  );
};

HeadCell.propTypes = {
  field: Type.string.isRequired,
  sortField: Type.string.isRequired,
  event: Type.func.isRequired,
};

export { UserList };
