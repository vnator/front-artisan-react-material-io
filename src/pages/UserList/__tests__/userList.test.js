import React from 'react';
import Enzyme from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import Adapter from 'enzyme-adapter-react-16';

import { UserList } from '../UserList';
import { mountWithIntl } from '../../../utils/enzymeHelper';
import { USER } from '../../../queries/user';

Enzyme.configure({ adapter: new Adapter() });

describe('UserList', () => {
  const mocks = [
    {
      request: {
        query: USER.GET_USER_LIST,
        variables: {
          sortField: '',
          skip: 0,
          limit: 10,
        },
      },
      result: {
        data: {
          users: [
            {
              id: 1,
              name: 'joao',
              email: 'joao@gmail.com',
              gender: 'MALE',
              dateOfBirth: '2011-02-02',
            },
            {
              id: 2,
              name: 'henrique',
              email: 'henrique@gmail.com',
              gender: 'MALE',
              dateOfBirth: '2000-02-02',
            },
            {
              id: 1,
              name: 'ana',
              email: 'ana@gmail.com',
              gender: 'FEMALE',
              dateOfBirth: '1997-02-02',
            },
          ],
        },
      },
    },
  ];

  it('Snapshot', () => {
    const userList = mountWithIntl(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <UserList />
        </MockedProvider>
      </MemoryRouter>,
    );

    expect(userList).toMatchSnapshot();
  });
});
