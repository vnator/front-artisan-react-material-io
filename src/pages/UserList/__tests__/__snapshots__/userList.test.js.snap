// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Snapshot 1`] = `ReactWrapper {}`;
